engines/scumm/dimuse_v2/dimuse_v2_waveout.cpp

```diff
@@ -0,0 +1,104 @@
+* of the License, or (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program; if not, write to the Free Software
+* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*
+*/
+
+#include "scumm/dimuse_v2/dimuse_v2.h"
+
+namespace Scumm {
+
+int DiMUSE_v2::waveOutInit(int sampleRate, waveOutParamsStruct *waveOutSettingsStruct) {
+	_waveOutSampleRate = sampleRate;
+
+	_waveOutBytesPerSample = 2;
+	_waveOutNumChannels = 2;
+	_waveOutZeroLevel = 0;
+	_waveOutPreferredFeedSize = 512;
+
+	// Nine buffers (waveOutPreferredFeedSize * 4 bytes each), one will be used for the mixer
+	_waveOutOutputBuffer = (uint8 *)malloc(_waveOutNumChannels * _waveOutBytesPerSample * _waveOutPreferredFeedSize * 9);
+	_waveOutMixBuffer = _waveOutOutputBuffer + (_waveOutNumChannels * _waveOutBytesPerSample * _waveOutPreferredFeedSize * 8); // 9-th buffer
+
+	// This information will be fed to the internal mixer during its initialization
+	waveOutSettingsStruct->bytesPerSample = _waveOutBytesPerSample * 8;
+	waveOutSettingsStruct->numChannels = _waveOutNumChannels;
+	waveOutSettingsStruct->mixBufSize = (_waveOutBytesPerSample * _waveOutNumChannels) * _waveOutPreferredFeedSize;
+	waveOutSettingsStruct->sizeSampleKB = 0;
+	waveOutSettingsStruct->mixBuf = _waveOutMixBuffer;
+
+	// Init the buffer filling it with zero volume samples
+	memset(_waveOutOutputBuffer, _waveOutZeroLevel, _waveOutNumChannels * _waveOutBytesPerSample * _waveOutPreferredFeedSize * 9);
+
+	_waveOutDisableWrite = 0;
+	return 0;
+}
+
+void DiMUSE_v2::waveOutWrite(uint8 **audioData, int *feedSize, int *sampleRate) {
+	uint8 *curBufferBlock;
+	if (_waveOutDisableWrite)
+		return;
+
+	if (_vm->_game.id == GID_DIG) {
+		_waveOutXorTrigger ^= 1;
+		if (!_waveOutXorTrigger)
+			return;
+	}
+	
+	*feedSize = 0;
+	if (_mixer->isReady()) {
+		curBufferBlock = &_waveOutOutputBuffer[_waveOutPreferredFeedSize * _waveOutWriteIndex * _waveOutBytesPerSample * _waveOutNumChannels];
+
+		*audioData = curBufferBlock;
+
+		*sampleRate = _waveOutSampleRate;
+		*feedSize = _waveOutPreferredFeedSize;
+		_waveOutWriteIndex = (_waveOutWriteIndex + 1) % 8;
+
+		byte *ptr = (byte *)malloc(_outputFeedSize * _waveOutBytesPerSample * _waveOutNumChannels);
+		memcpy(ptr, curBufferBlock, _outputFeedSize * _waveOutBytesPerSample * _waveOutNumChannels);
+
+		_internalMixer->_stream->queueBuffer(ptr,
+			_outputFeedSize * _waveOutBytesPerSample * _waveOutNumChannels,
+			DisposeAfterUse::YES,
+			Audio::FLAG_16BITS | Audio::FLAG_STEREO | Audio::FLAG_LITTLE_ENDIAN);
+	}
+}
+
+int DiMUSE_v2::waveOutDeinit() {
+	_waveOutDisableWrite = 1;
+	free(_waveOutOutputBuffer);
```

Comment:
should get reset to `NULL` to prevent double frees - and maybe should also get freed in the dtor of the class (this is also true for other buffers)

Reply:
Thanks, this one actually made me realize that dispatchDeinit wasn't called anywhere, so I never freed _dispatchBuffer! :P I guess this is a very good reason to move all these free() ops in the dtor.